### PR TITLE
Fix #240 Update SAB Branch

### LINKED ISSUE #240: Update SAB branch to v1.0
Rebase the `feature/SABdraft` onto the latest `master` and fix up based on changes. However, evaluate if that is the best approach given that so much has changed.

Note what needs to be done to finish this feature also [here](https://docs.google.com/document/d/13UMhnE6oGXZfZ0hK9XD56NG1-jaDeCxq4S3GTqG-qfo/edit?usp=sharing).
Sample SAB data is [here](https://drive.google.com/drive/folders/1a2Ggia3GbIFgFiB9GwCmoF6TLOV9-HI0?usp=sharing).

### PR DESCRIPTION
**Pull request by JCFarrow**
BK can now import and render with SAB projects

### CODE REVIEW

File: main/sources/scripture-app-builder.ts [hunk 1]
```diff
@@ -1,18 +1,286 @@
...
+    let timingFileContent;
+
+    try {
+      timingFileContent = fs.readFileSync(timingFileFullName, 'utf8');
+    } catch (error) {
+      console.error(error);
+      return [];
+    }
+
+    const timingPattern = /(\d+\.?\d*)\t(\d+\.?\d*)\t(\d+)/gs;
```
Comment (irahopkinson) [inline main/sources/scripture-app-builder.ts:222]: RegEx are tricky things. It is preferable to add a comment stating what the intended purpose/result is.

File: main/sources/scripture-app-builder.ts [hunk 2]
```diff
@@ -1,18 +1,286 @@
...
+
+    let timingMatches;
+    const timingOutput = [];
+    while ((timingMatches = timingPattern.exec(timingFileContent))) {
+      // console.log(timingMatches[1] + '..' + timingMatches[2] + '...' + timingMatches[3]);
```
Comment (irahopkinson) [inline main/sources/scripture-app-builder.ts:227]: This looks like a debug line. Is it still needed?

File: main/sources/scripture-app-builder.ts [hunk 3]
```diff
@@ -1,18 +1,286 @@
...
+        verse: timingMatches[3],
+      });
+    }
+
+    const chapterJson = sfmJson.chapters ? sfmJson.chapters[chapterNumber] : undefined;
+    if (!chapterJson) {
+      return [];
+    }
+    for (const verse in chapterJson) {
+      const startTime = _.find(timingOutput, { verse: verse })?.startTime;
```
Comment (irahopkinson) [inline main/sources/scripture-app-builder.ts:241]: TypeScript has a simpler syntax when the variable name matches the property name. Change `{ verse: verse }` to `{ verse }`. Here and below also.

**Review by irahopkinson — CHANGES_REQUESTED** [review]
Great job with finishing this feature off. A few things to fix or consider changing.

File: src/App/store/Settings.ts [hunk 4]
```diff
@@ -22,7 +22,7 @@ export const getDefaultScriptureAppBuilderDirectory = (): string => {
       return path.join(os.homedir(), 'Documents', 'App Builder', 'Scripture Apps', 'App Projects');
     case 'darwin':
     default:
-      return `${os.homedir()}/App Builder/Scripture Apps/App Projects/`;
+      return `${os.homedir()}/Documents/AppBuilder/Scripture Apps/App Projects/`;
```
[the remark this answers is not in the document]
Comment (JCFarrow) [reply src/App/store/Settings.ts:25]: Yes, windows has a space, see the above win32 case line. This line is for mac I think and I don't have one so I can't check it. This change was made by Ray, it's possible that he missed the space?

[the remark this answers is not in the document]
Comment (irahopkinson) [reply src/App/store/Settings.ts:25]: Oops, was thinking that was the Windows line. Right you are. I can't fully remember if he was working on a Mac but I think so. So lets leave this as is. We will need to check this when we do a Mac build.

File: main/sources/scripture-app-builder.ts [hunk 5]
```diff
@@ -1,18 +1,286 @@
...
   }
+
+  private makeProject(projectName: string, directory: string): BKProject | BKProject[] {
+    const emptyProject = {
+      name: projectName,
+      folderPath: path.join(directory, projectName),
+      sourceType: this.SOURCE_TYPE,
+      books: [],
+    };
+    const fileName = path.join(directory, projectName, projectName) + '.appDef';
```
[the remark this answers is not in the document]
Comment (JCFarrow) [reply main/sources/scripture-app-builder.ts:59]: This is how I found the code. In all my test projects the project name is the same between the main folder, the _data folder and the .appDef file. It's possible that the main folder could get renamed or something. Maybe we should check just for the endings (".appDef" and "_data").
I just tested with SAB and changing the project name within SAB changes the name in all three places. It looks like the only way to mess it up is if someone manually renames the folder. Should we be allowing for that?

File: main/sources/scripture-app-builder.ts (hunk 3 above)
Comment (JCFarrow) [reply main/sources/scripture-app-builder.ts]: done

File: main/sources/scripture-app-builder.ts (hunk 2 above)
Comment (JCFarrow) [reply main/sources/scripture-app-builder.ts]: removed

File: main/sources/scripture-app-builder.ts (hunk 1 above)
Comment (JCFarrow) [reply main/sources/scripture-app-builder.ts:226]: added some clarification

File: main/sources/scripture-app-builder.ts [hunk 6]
```diff
@@ -1,18 +1,286 @@
...
+      const project: BKProject = {
+        name: projectName,
+        folderPath: path.join(directory, projectName),
+        sourceType: this.SOURCE_TYPE,
+        books: this.makeBooks(xmlDoc, projectName, directory, collectionIds[0]),
+      };
+      return project;
+    } else if (collectionIds.length > 1) {
+      const projects = collectionIds.map((id) => {
+        const project = {
```
[the remark this answers is not in the document]
Comment (JCFarrow) [reply main/sources/scripture-app-builder.ts]: Good idea. Done

File: main/sources/scripture-app-builder.ts [hunk 7]
```diff
@@ -1,18 +1,286 @@
...
+      return [];
+    }
   }
+
   reloadProject(project: BKProject): BKProject {
     return project;
   }
+
+  private makeProject(projectName: string, directory: string): BKProject | BKProject[] {
+    const emptyProject = {
```
[the remark this answers is not in the document]
Comment (JCFarrow) [reply main/sources/scripture-app-builder.ts]: done

File: main/sources/scripture-app-builder.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (JCFarrow) [reply main/sources/scripture-app-builder.ts]: done

File: main/sources/scripture-app-builder.ts (hunk 1 above)
Comment (irahopkinson) [reply main/sources/scripture-app-builder.ts:226]: Nice

File: main/sources/scripture-app-builder.ts (hunk 3 above)
Comment (irahopkinson) [reply main/sources/scripture-app-builder.ts]: Good. There is more below in `segment.push`...

File: main/sources/scripture-app-builder.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (irahopkinson) [reply main/sources/scripture-app-builder.ts]: Thanks for checking in SAB. That's enough for me.

File: main/sources/scripture-app-builder.ts (hunk 3 above)
Comment (JCFarrow) [reply main/sources/scripture-app-builder.ts]: Now it's done.

**Review by irahopkinson — APPROVED** [review]
👍 LGTM. Thanks for all your work on this, good job.

**Pull request merged**